### PR TITLE
feat(row-based encoding): impl Encoding and Decoding for row-based

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Now the trait `Encoding` and `Decoding` is heavily bound to cell-based format, however, they should be more normal and not aware of specific encoding format.  This PR refactor these two trait, and the following PR will do such things:
- merge `Encoding` and `Decoding` into one trait to be used by storage table. (just as an example)
```
trait Encoding {
  fn serialize(&self, ..)
}

trait Decoding {
  fn deserialize(&mut self, ..) 
}

trait Xxx {
  type Serializer: Encoding
  type Deserializer: Decoding

  fn create_serializer(..) -> Self::Serializer;
  fn create_deserializer(..) -> Self::Deserializer;
}

struct StorageTableBase<S: StateStore, X: Xxx>
```
- implement Row-based StateTable. (just as an example).
```
pub type StateTable<S> = StateTableBase<S, CellBasedRowSerializer>;
pub type RowBasedStateTable<S> = StateTableBase<S, RowBasedSerializer>;
```

As you can see, some of the parameters passed are not used yet, because current row-based encoding dose not implement PK dudup, but I think this is needed and will be done in the future. 
## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/storage/src/encoding/row_based_serializer.rs [hunk 1]
```diff
@@ -14,30 +14,51 @@
...
+    fn create_row_serializer(
+        _pk_indices: &[usize],
+        _column_descs: &[risingwave_common::catalog::ColumnDesc],
+        _column_ids: &[risingwave_common::catalog::ColumnId],
+    ) -> Self {
+        // todo: Change row-based serializer with dedup pk
+        Self {}
+    }
+
+    fn cell_based_serialize_without_filter(
```
Comment (wcy-fdu) [inline src/storage/src/encoding/row_based_serializer.rs:52]: Need to rename this fn to something else, removing the cell-based attribute.

File: src/storage/src/encoding/row_based_deserializer.rs [hunk 2]
```diff
@@ -18,67 +18,57 @@ use risingwave_common::error::Result;
...
+        //todo: raw_key will be used in row-based pk dudup later.
+        Ok(Some((
+            0,
+            raw_key.as_ref().to_vec(),
+            row_based_deserialize_inner(self.data_types.clone(), value.as_ref())?,
+        )))
     }
 }
 
+pub fn row_based_deserialize_inner(data_types: Vec<DataType>, mut row: impl Buf) -> Result<Row> {
```
Comment (BugenZhao) [inline src/storage/src/encoding/row_based_deserializer.rs:54]: No need `pub` here?

**Review by BugenZhao — COMMENTED** [review]
Could you please also refactor the `get` with `Decoding` trait to check whether this interface fits our usages?

File: src/storage/src/encoding/row_based_serializer.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/storage/src/encoding/row_based_serializer.rs]: LGTM

**Comment by wcy-fdu** [comment]
> Could you please also refactor the `get` with `Decoding` trait to check whether this interface fits our usages?

Sure, let me try it.

**Comment by wcy-fdu** [comment]
> > Could you please also refactor the `get` with `Decoding` trait to check whether this interface fits our usages?
> 
> Sure, let me try it.

Maybe better to refactor it after merging `Encoding` and `Decoding` into one trait?Otherwise there will be two trait `E` and `D` in StorageTable, not elegant enough ...

**Comment by BugenZhao** [comment]
> > > Could you please also refactor the `get` with `Decoding` trait to check whether this interface fits our usages?
> > 
> > 
> > Sure, let me try it.
> 
> Maybe better to refactor it after merging `Encoding` and `Decoding` into one trait?Otherwise there will be two trait `E` and `D` in StorageTable, not elegant enough ...

LGTM. Seems we're already using the functions defined by the trait in `get` besides `new`.

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**